src/mpid/common/shm/mpidu_init_shm_alloc.c

```diff
@@ -0,0 +1,652 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+#include <mpidimpl.h>
+#include "mpl_shm.h"
+
+#include <stdlib.h>
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+#include <errno.h>
+
+#if defined (HAVE_SYSV_SHARED_MEM)
+#include <sys/ipc.h>
+#include <sys/shm.h>
+#endif
+
+#if defined(HAVE_MKSTEMP) && defined(NEEDS_MKSTEMP_DECL)
+extern int mkstemp(char *t);
+#endif
+
+#include "mpidu_init_shm_alloc.h"
+#include "mpidu_shm_impl.h"
+#include "mpidu_generic_queue.h"
+
+typedef struct alloc_elem {
+    struct alloc_elem *next;
+    void **ptr_p;
+    size_t len;
+} alloc_elem_t;
+
+static struct {
+    alloc_elem_t *head, *tail;
+} allocq = {
+0};
+
+static int check_alloc(MPIDU_shm_seg_t * memory, MPIDU_shm_barrier_t * barrier,
+                       int num_local, int local_rank);
+
+#define ALLOCQ_HEAD() GENERIC_Q_HEAD(allocq)
+#define ALLOCQ_EMPTY() GENERIC_Q_EMPTY(allocq)
+#define ALLOCQ_ENQUEUE(ep) GENERIC_Q_ENQUEUE(&allocq, ep, next)
+#define ALLOCQ_DEQUEUE(epp) GENERIC_Q_DEQUEUE(&allocq, epp, next)
+
+static size_t segment_len = 0;
+
+static int num_segments = 0;
+
+typedef struct asym_check_region {
+    void *base_ptr;
+    OPA_int_t is_asym;
+} asym_check_region;
+
+static asym_check_region *asym_check_region_p = NULL;
+
+/* MPIDU_Init_shm_seg_alloc(len, ptr_p)
+
+   This function is used to allow the caller to reserve a len sized
+   region in the shared memory segment.  Once the shared memory
+   segment is actually allocated, when MPIDU_Init_SHM_Seg_commit() is
+   called, the pointer *ptr_p will be set to point to the reserved
+   region in the shared memory segment.
+
+   Note that no shared memory is actually allocated by this function,
+   and the *ptr_p pointer will be valid only after
+   MPIDU_Init_SHM_Seg_commit() is called.
+*/
+int MPIDU_Init_shm_seg_alloc(size_t len, void **ptr_p, MPL_memory_class class)
+{
```

Comment:
Do we need this function? For example, can we simply use a predefine segment size?

Reply:
Yes I plan to refactor this one too and get rid of most of the complexity